benches/mock_bench/internal_pointer.rs

```diff
@@ -0,0 +1,97 @@
+use criterion::Criterion;
+
+#[cfg(feature = "is_mmtk_object")]
+use mmtk::util::test_util::fixtures::*;
+use mmtk::util::test_util::mock_method::*;
+use mmtk::util::test_util::mock_vm::{write_mockvm, MockVM};
+
+pub fn bench(c: &mut Criterion) {
+    // Setting a larger heap, although the GC should be disabled in the MockVM
+    #[cfg(feature = "is_mmtk_object")]
+    let mut fixture = MutatorFixture::create_with_heapsize(1 << 30);
+
+    // Normal objects
+    // 16KB object -- we want to make sure the object can fit into any normal space (e.g. immix space or mark sweep space)
+    const NORMAL_OBJECT_SIZE: usize = 16 * 1024;
+    write_mockvm(|mock| {
+        *mock = MockVM {
+            get_object_size: MockMethod::new_fixed(Box::new(|_| NORMAL_OBJECT_SIZE)),
+            is_collection_enabled: MockMethod::new_fixed(Box::new(|_| false)),
+            ..MockVM::default()
+        }
+    });
+
+    c.bench_function("internal pointer - normal objects", |_b| {
```

Comment:
This is some data for this micro benchmark.

The naive implementation (`find_prev_non_zero_value_simple`, similar to https://github.com/mmtk/mmtk-core/pull/1155): check every aligned data address until we find the object.
```console
internal pointer - normal objects
                        time:   [12.910 µs 12.911 µs 12.912 µs]

internal pointer - large objects
                        time:   [53.164 ns 53.486 ns 53.939 ns]
```

Check every _byte_ in side metadata until we find VO bit, and compute it back to the VO address
```console
internal pointer - normal objects
                        time:   [1.5500 µs 1.5504 µs 1.5509 µs]
                        change: [-88.001% -87.997% -87.994%] (p = 0.00 < 0.05)
                        Performance has improved.

internal pointer - large objects
                        time:   [58.115 ns 58.137 ns 58.161 ns]
                        change: [+8.0877% +8.6753% +9.1317%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

Check every _word_ in side metadata until we find VO bit, and compute it back to the VO address
```console
internal pointer - normal objects
                        time:   [311.28 ns 312.38 ns 313.51 ns]
                        change: [-79.856% -79.804% -79.741%] (p = 0.00 < 0.05)
                        Performance has improved.

internal pointer - large objects
                        time:   [57.582 ns 57.607 ns 57.635 ns]
                        change: [-0.9006% -0.8311% -0.7566%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```

Reply:
After applying the suggestion in https://github.com/mmtk/mmtk-core/pull/1165#discussion_r1671931368, we only check if the metadata address is mapped for every chunk. The performance is further improved.
```console
internal pointer - normal objects
                        time:   [139.48 ns 139.65 ns 139.82 ns]
                        change: [-55.524% -55.388% -55.250%] (p = 0.00 < 0.05)
                        Performance has improved.

internal pointer - large objects
                        time:   [49.300 ns 49.361 ns 49.462 ns]
                        change: [-14.459% -14.367% -14.263%] (p = 0.00 < 0.05)
                        Performance has improved.
```